Single-file diagnostic support for SOS.

mikem8361 opened the pull request:
This will allow SOS, dotnet-dump, etc. to work for single-file programs on Linux. An library called runtimeinfo_lib is generated that is to be linked into the single-file superhost containing coreclr.

Added "DotNetRuntimeInfo" export/block containing the symbol index/build ids needed by the diagnostic tools.

Added ICLRRuntimeLocator DAC interface for SOS and other DAC clients so their data targets can provide the single-file runtime base address.

Added the GetModuleIndex tool to get build ids.

PAL fix for single-file superhost.

Commit:
Single-file diagnostic support.

This will allow SOS, dotnet-dump, etc. to work for single-file programs on Linux. An library called runtimeinfo_lib is generated that is to be linked into the single-file superhost containing coreclr.

Added "DotNetRuntimeInfo" export/block containing the symbol index/build ids needed by the diagnostic tools.

Added ICLRRuntimeLocator DAC interface for SOS and other DAC clients so their data targets can provide the single-file runtime base address.

Added the GetModuleIndex tool to get build ids.

PAL fix for single-file superhost.

Commit:
Fix elfreader allocation assert in debug. The 'new' did not use utilcode overrides but the 'delete' did.

mikem8361 commented:
The SOS PR that uses this support: https://github.com/dotnet/diagnostics/pull/998

lpereira commented on src/coreclr/src/debug/dbgutil/elfreader.cpp:
```diff
@@ -237,7 +237,7 @@ ElfReader::InitializeGnuHashTable()
         return false;
     }
-    m_buckets = new (std::nothrow) int32_t[m_hashTable.BucketCount];
+    m_buckets = (int32_t*)malloc(m_hashTable.BucketCount * sizeof(int32_t));
```
Any reason to switch from `new[]` to `malloc()`?

mikem8361 replied:
The runtime has these overrides that replace new/delete with ones in clrhost_nodependencies.cpp but for the elfreader, the "new" wasn't going through those overrides but the "delete" was causing an assert in the debug build. It was simpler to use malloc/free.

lpereira commented on src/coreclr/src/pal/prebuilt/idl/clrdata_i.cpp:
```diff
@@ -1,15 +1,21 @@
-// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
-
```
Shouldn't the license goo be retained?

lpereira replied:
(`git add -p` is your friend!)

sdmaclea replied:
I like to spell it out `git add --patch` or `git commit --patch`.  It iterates through changes so you can select which to add/commit.

mikem8361 replied:
I add back the header.

am11 replied:
also `git add --interactive` or `-i`.

> It iterates through changes

Those changes are called hunks in [libxdiff](https://github.com/git/git/tree/master/xdiff) jargon. :)

lpereira commented on src/coreclr/src/pal/src/loader/module.cpp:
```diff
@@ -963,6 +963,8 @@ BOOL LOADInitializeModules()
     exe_module.prev = &exe_module;
     exe_module.pDllMain = nullptr;
     exe_module.hinstance = nullptr;
+    exe_module.pDllMain = (PDLLMAIN)dlsym(exe_module.dl_handle, "DllMain");
```
```suggestion
    exe_module.pDllMain = (PDLLMAIN)dlsym(exe_module.dl_handle, "DllMain");
    if (exe_module.pDllMain == nullptr)
    {
        ERROR("Boom!\n");
        return FALSE;
    }
```

mikem8361 replied:
That would cause the normal "hosts" (dotnet, corerun, etc.) to fail because they don't have a DllMain.

lpereira commented on src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt:
```diff
@@ -164,6 +164,29 @@ endif(FEATURE_EVENT_TRACE)
 
+# Create the runtime module index header file containing the coreclr build id
+# for xplat and the timestamp/size on Windows.
+if(CLR_CMAKE_HOST_LINUX OR CLR_CMAKE_HOST_WIN32)
```
Do we have a `CLR_CMAKE_HOST_ELF_PLATFORM` (or something), or do we really rely on Linux-only syscalls to make this thing work?

mikem8361 replied:
No there isn't. This is a kind of a special condition of both Windows and Linux (but not MacOS and the FreeBSD). It wouldn't be worth it to add a special flag.

jkoritzinsky reviewed (changes requested):
The `src/coreclr/build.cmd` and `src/coreclr/build.sh` files will be going away soon (once I move the formatting jobs off from using them). Can you add the `GetModuleIndex.csproj` project to the `clr.tools` subset in `eng/Subsets.props`?

mikem8361 commented:
@jkoritzinsky, yes those build script changes are kind of a hack and it sounds like what you suggest is a lot cleaner and long as the GetModuleIndex project gets built before the native build.

jkoritzinsky commented:
If it needs to be built before the native build, then we'd need to make a new subset. Which is fine, since we'd also move DacTableGen into that subset. We could name it something like `clr.diagtools`.

mikem8361 commented:
How do I add a new subset?

jkoritzinsky commented:
Here's a few steps for adding a new subset (all in `eng/Subsets.props`):

1. Update the `DefaultCoreClrSubsets` property to include your new subset (separated by `+`)
2. Add a `SubsetName` item that describes the new subset.
3. At line 146, add this block:

```
  <ItemGroup Condition="$(_subset.Contains('+clr.diagtools+'))">
    <CoreClrProject Include="$(CoreClrProjectRoot)src\tools\GetModuleIndex\GetModuleIndex.csproj" />
  </ItemGroup>
```

Commit:
Code review feedback.

jkoritzinsky commented on eng/Subsets.props:
```diff
@@ -179,6 +180,10 @@
     <CoreClrProject Include="$(CoreClrProjectRoot)src\tools\crossgen2\crossgen2\crossgen2.csproj" />
   </ItemGroup>
 
+  <ItemGroup Condition="$(_subset.Contains('-diagtools-'))">
```
This block needs to be moved above the block at line 159 (the line reference I gave was based on master).

Commit:
Fix GetModuleIndex building and execution

am11 commented on src/coreclr/CMakeLists.txt:
```diff
@@ -25,6 +25,14 @@ set(GENERATED_EVENTING_DIR ${CMAKE_CURRENT_BINARY_DIR}/Eventing)
 
+if(CLR_CMAKE_HOST_UNIX)
+    set(CLR_DOTNET_COMMAND ${CLR_REPO_ROOT_DIR}/dotnet.sh)
+endif(CLR_CMAKE_HOST_UNIX)
+
+if(CLR_CMAKE_HOST_WIN32)
+    set(CLR_DOTNET_COMMAND ${CLR_REPO_ROOT_DIR}/dotnet.cmd)
+endif(CLR_CMAKE_HOST_WIN32)
```
nit:
```suggestion
elseif(CLR_CMAKE_HOST_WIN32)
    set(CLR_DOTNET_COMMAND ${CLR_REPO_ROOT_DIR}/dotnet.cmd)
endif(CLR_CMAKE_HOST_UNIX)
```

Commit:
Needed to add a line to the build-job.yml pipeline to build the new clr.diagtools subset before the native coreclr runtime build because it uses the build-runtime scripts directly.

Added a Versions.props property for the Microsoft.FileFormats version and updated it to the latest.

Changed the GetModuleIndex to take an input and output file instead of using output redirection to write it because some messages from the dotnet.cmd script used to run it got written to the output h file.

Commit:
Fix x86, arm builds

am11 commented:
With current master:

```sh
# coreclr native compilation without other stuff
cd runtime
git clean -xdf
./src/coreclr/build-runtime.sh -skipgenerateversion -nopgooptimize
```

is the fastest way to compile coreclr that skips msbuild and does not restore anything. PR forces it to download dotnet(1).

Can we decouple dotnet dependency from cmake so `CLR_DOTNET_COMMAND` and custom command in cmake are not needed? maybe another subset which runs after native compilation finishes?

am11 commented on src/coreclr/src/pal/prebuilt/inc/clrdata.h:
```diff
@@ -1,13 +1,19 @@
-// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
-
```
nit: revert header in this file as well.

janvorli commented on src/coreclr/src/debug/runtimeinfo/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+
+add_library_clr(runtimeinfo_lib STATIC ${RUNTIMEINFO_SOURCES})
+
+add_dependencies(runtimeinfo_lib runtime_module_index_header)
```
I would prefer dropping the `_lib` suffix. We don't have any other library with such a naming convention.

janvorli commented on src/coreclr/src/dlls/mscordbi/CMakeLists.txt:
```diff
@@ -117,5 +117,28 @@ elseif(CLR_CMAKE_HOST_UNIX)
+    add_custom_command(
+        OUTPUT ${DBI_MODULE_INDEX_FILE}
+        COMMAND ${CLR_DOTNET_COMMAND} ${CMAKE_INSTALL_PREFIX}/GetModuleIndex/GetModuleIndex.dll $<TARGET_FILE:mscordbi> ${DBI_MODULE_INDEX_FILE}
+        DEPENDS $<TARGET_FILE:mscordbi> 
```
A nit - you could use just `DEPENDS mscordbi` here.

janvorli commented on src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt:
```diff
@@ -164,6 +164,29 @@ endif(FEATURE_EVENT_TRACE)
+    add_custom_command(
+        OUTPUT ${RUNTIME_MODULE_INDEX_FILE}
+        COMMAND ${CLR_DOTNET_COMMAND} ${CMAKE_INSTALL_PREFIX}/GetModuleIndex/GetModuleIndex.dll $<TARGET_FILE:coreclr> ${RUNTIME_MODULE_INDEX_FILE}
+        DEPENDS $<TARGET_FILE:coreclr> 
```
A nit - you could use just `DEPENDS coreclr` here.

Commit:
More code review feedback

janvorli reviewed (approved):
LGTM, thank you!

swaroop-sridhar commented on src/coreclr/src/debug/CMakeLists.txt:
```diff
@@ -3,3 +3,6 @@ add_subdirectory(ildbsymlib)
 add_subdirectory(ee)
 add_subdirectory(di)
 add_subdirectory(shim)
+if(CLR_CMAKE_HOST_LINUX OR CLR_CMAKE_HOST_WIN32)
```
Why is this only Linux and Windows?

mikem8361 replied:
I didn't think we will ever support single-file on MacOS. Now as far as building the runtimeinfo library, I could make some changes to GetModuleIndex and the build system to support it if it important.

swaroop-sridhar replied:
Publishing an app as a single-file is currently supported on MacOS, but it is not a very interesting case, and there are no known users.

In the short term, we'll continue to support single-file apps on MacOS, but there are no plans to build the super-host (that is, coreclr will be a separate file on disk beside the app).  

So, the change is fine as-is. I was just curious if MacOS was excluded because we don't need the library, or if there were any other challenges. Thanks.

swaroop-sridhar reviewed (approved):
Great to see have this work @mikem8361. Thanks.
Since we don't have the single-file super-host, how did you test the changes?

mikem8361 commented:
I made some changes to corerun to make it a superhost as far as linking just coreclr into it. The changes are in a separate branch in my fork:

https://github.com/mikem8361/runtime/tree/sfcorerun

mikem8361 commented:
@jkoritzinsky do you have anymore feedback?

jkoritzinsky commented on eng/pipelines/coreclr/templates/build-job.yml:
```diff
@@ -124,6 +124,10 @@ jobs:
           installDotnet: true
 
+    # Build CoreCLR Diagnostic tools needed by the native runtime build
+    - script: $(Build.SourcesDirectory)$(dir)build$(scriptExt) -subset clr.diagtools $(crossArg) -arch $(archType) -c $(buildConfig) $(officialBuildIdArg) -ci
```
I think we need to pass the -bl flag with a custom name so we generate a binlog for this step.

am11 commented:
@mikem8361, could we add a `FEATURE_SINGLE_FILE_DIAGNOSTICS` in https://github.com/dotnet/runtime/blob/master/src/coreclr/clrfeatures.cmake and move `if(CLR_CMAKE_HOST_LINUX OR CLR_CMAKE_HOST_WIN32)` condition there, so we have some way to turn it off for faster builds (https://github.com/dotnet/runtime/pull/34697#issuecomment-611354395)? This is the same way we enable/disable other features like event trace etc.

jkoritzinsky reviewed (commented):
One last thing, but other than that it all looks good!

mikem8361 commented:
@am11, even if we had a FEATURE_SINGLE_FILE_DIAGNOSTICS flag, it would be defined by default. Not sure of the scenario where it would be off especially when start on the single-file "superhost" work.

Commit:
Code review feedback

am11 commented:
@mikem8361, most of the features (`FEATURE_DBGIPC`, `FEATURE_EVENT_TRACE`) in clrfeatures are enabled by default for CI builds, but the value is that it gives a control to developer with `build-runtime.sh -cmakeargs -DFEATURE_XXX=OFF`, to disable certain feature while validating particular scenario.

Moving the condition to clrfeatures.cmake will prevent the regression being introduced by the PR.

noahfalk commented on eng/Subsets.props:
```diff
@@ -80,6 +80,7 @@
     <SubsetName Include="Clr.CoreLib" Description="The managed System.Private.CoreLib library for CoreCLR." />
     <SubsetName Include="Clr.NativeCoreLib" Description="Run crossgen on System.Private.CoreLib library for CoreCLR." />
     <SubsetName Include="Clr.Tools" Description="Managed tools that support CoreCLR development and testing." />
+    <SubsetName Include="Clr.DiagTools" Description="Diagnostics tools needed to build CoreCLR." />
```
What do you think of?
```suggestion
    <SubsetName Include="Clr.BuildTools" Description="Diagnostics tools needed to build CoreCLR." />
```

I am guessing the high order bit for the developers working in clr is that you now need this tool in order to build the runtime, regardless whether you care about using the diagnostics or not.

sdmaclea replied:
If that were the case I would eliminate the new subset and add it to the `Clr.Tools` subset.

jkoritzinsky replied:
The Clr.Tools subset has to be built after CoreCLR because crossgen2 need the native build to have completed.

noahfalk replied:
The Tools subset aren't required for building as I understood it. I'd imagine that is an important distinction if you are trying to build the minimal amount necessary to produce the runtime.

sdmaclea replied:
I have been thinking we should have prerequisite subsets.

mikem8361 replied:
I do like changing the new subset to "BuildTools"  because it does describe it more generally.  Eventually building DacTableGen should be moved to this subset like Jeremy suggested.

I'll rename it.

noahfalk commented on src/coreclr/src/pal/src/loader/module.cpp:
```diff
@@ -963,6 +963,8 @@ BOOL LOADInitializeModules()
     exe_module.prev = &exe_module;
     exe_module.pDllMain = nullptr;
```
Nit: These two lines appear unnecessary now.

noahfalk reviewed (approved):
Thanks Mike, LGTM!

Commit:
More code review feedback

Commit:
Add FEATURE_SINGLE_FILE_DIAGNOSTICS cmake flag so build-runtime.sh can work without clr.buildtools subset
